### PR TITLE
support add-on configuration apis

### PR DESCRIPTION
**Pull request by skeeey**
Signed-off-by: Wei Liu <liuweixa@redhat.com>

### CODE REVIEW

File: pkg/addonfactory/addonfactory.go [hunk 1]
```diff
@@ -145,3 +160,20 @@ func (f *AgentAddonFactory) BuildTemplateAgentAddon() (agent.AgentAddon, error)
...
+		// config gvr is not defined, ignore
+		return nil
+	}
+
+	if configGVR.Version == "" {
+		return fmt.Errorf("config version is required")
+	}
+
+	if configGVR.Resource == "" {
+		return fmt.Errorf("config resoruce is required")
```
Comment (thomasmckay) [inline pkg/addonfactory/addonfactory.go]: nitpick: typo `resource`

Comment (skeeey) [reply pkg/addonfactory/addonfactory.go]: yeah, fixed, thank you

File: pkg/agent/inteface.go [hunk 2]
```diff
@@ -65,6 +70,9 @@ type AgentAddonOptions struct {
 	// If not set, will be defaulted to false.
 	// +optional
 	HostedModeEnabled bool
+
+	// SupportedConfigGVRs is an addon configuration GroupVersionResource set that is supported by current addon.
+	SupportedConfigGVRs map[schema.GroupVersionResource]EmptyObjectType
```
Comment (qiujian16) [inline pkg/agent/inteface.go]: shoudn't this be runtime.Object? or if value is not necessary, can it be a list?

File: pkg/addonmanager/controllers/addonconfig/controller.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+		WithInformersQueueKeysFunc(func(obj runtime.Object) []string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return []string{key}
+		}, addonInformers.Informer()).
+		WithBareInformers(configInformer.Informer()).
+		WithSync(c.sync).ToController(controllerName)
+}
+
+func (c *addonConfigController) enqueueAddOnsByConfig(obj interface{}) {
+	name, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
```
Comment (qiujian16) [inline pkg/addonmanager/controllers/addonconfig/controller.go]: I am not sure name is sufficient, I think you need gvr also?

File: pkg/addonmanager/manager.go [hunk 4]
```diff
@@ -190,12 +204,24 @@ func (a *addonManager) Start(ctx context.Context) error {
...
 	go clusterInformers.Start(ctx.Done())
 	go kubeInfomers.Start(ctx.Done())
+	go dynamicInformers.Start(ctx.Done())
 
 	go deployController.Run(ctx, 1)
 	go registrationController.Run(ctx, 1)
 	go clusterManagementController.Run(ctx, 1)
 	go addonInstallController.Run(ctx, 1)
 	go addonHealthCheckController.Run(ctx, 1)
+	for _, configGVR := range a.addonConfigs {
```
Comment (qiujian16) [inline pkg/addonmanager/manager.go]: what about we start a single controller that watches multiple gvrs?

File: pkg/addonmanager/controllers/clustermanagement/controller.go [hunk 5]
```diff
@@ -95,42 +96,39 @@ func (c *clusterManagementController) sync(ctx context.Context, syncCtx factory.
...
 		Resource: "clustermanagementaddons",
 		Group:    addonapiv1alpha1.GroupVersion.Group,
 	})
 
+	// Add config references
+	if err := mergeConfigReferences(&modified, c.agentAddons[addonName], clusterManagementAddon, addonCopy); err != nil {
+		meta.SetStatusCondition(&addonCopy.Status.Conditions, metav1.Condition{
+			Type:    addonapiv1alpha1.ManagedClusterAddOnCondtionConfigured,
+			Status:  metav1.ConditionFalse,
+			Reason:  "ConfigurationWrong",
```
Comment (qiujian16) [inline pkg/addonmanager/controllers/clustermanagement/controller.go]: UnSupportedConfiguration, when do we set it true

File: pkg/agent/inteface.go (hunk 2 above)
Comment (skeeey) [reply pkg/agent/inteface.go]: change to schema.GroupVersionResource list, if using runtime.Object, it seems that it's not easy to get the type of object

File: pkg/addonmanager/controllers/addonconfig/controller.go (hunk 3 above)
Comment (skeeey) [reply pkg/addonmanager/controllers/addonconfig/controller.go]: change to one controller to handle all of configs

File: pkg/addonmanager/manager.go (hunk 4 above)
Comment (skeeey) [reply pkg/addonmanager/manager.go]: using a single controller to do this

File: pkg/addonmanager/controllers/clustermanagement/controller.go (hunk 5 above)
Comment (skeeey) [reply pkg/addonmanager/controllers/clustermanagement/controller.go]: changed

File: pkg/addonmanager/controllers/addonconfig/controller.go [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+
+		if errors.IsNotFound(err) {
+			continue
+		}
+
+		if err != nil {
+			return err
+		}
+
+		generation, err := getConfigGeneration(config)
```
Comment (qiujian16) [inline pkg/addonmanager/controllers/addonconfig/controller.go]: if config is unstructured, you can directly get generation. something to note is configmap and secret will not increment generation, we should add a TODO here.

File: pkg/addonmanager/controllers/addonconfig/controller.go [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+				continue
+			}
+
+			if configReference.Group != gvr.Group || configReference.Resource != gvr.Resource {
+				// uninterested config, ignore
+				continue
+			}
+
+			if configReference.Namespace != "" {
+				configNames = append(configNames, configReference.Namespace+"/"+configReference.Name)
```
Comment (qiujian16) [inline pkg/addonmanager/controllers/addonconfig/controller.go]: I think you also need gvr in the key.

File: pkg/addonmanager/manager.go [hunk 8]
```diff
@@ -153,6 +167,13 @@ func (a *addonManager) Start(ctx context.Context) error {
 		a.addonAgents,
 	)
 
+	addonConfigController := addonconfig.NewAddonConfigController(
```
Comment (qiujian16) [inline pkg/addonmanager/manager.go]: nit: let's check if there is configs and start this part. I think we should have a separated start func  to start it.

File: pkg/addonmanager/controllers/addonconfig/controller.go (hunk 6 above)
Comment (skeeey) [reply pkg/addonmanager/controllers/addonconfig/controller.go]: get the generation directly and added the TODO

File: pkg/addonmanager/manager.go (hunk 8 above)
Comment (skeeey) [reply pkg/addonmanager/manager.go]: checked

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**